tests/framework/validators/ValidatorTest.php

```diff
@@ -214,4 +214,11 @@ public function testAddError()
         $errors = $m->getErrors('attr_msg_val');
         $this->assertEquals('attr_msg_val::abc::param_value', $errors[0]);
     }
+
+    public function testGetActiveValidatorsForSafeAttributes()
+    {
+        $model = $this->getTestModel();
+        $validators = $model->getActiveValidators('safe_attr');
+        $this->assertInstanceOf('yii\validators\NumberValidator', array_shift($validators));
```

Comment:
ahh, you are right. That won't work because the keys are not kept in sync. Related issue topic is https://github.com/yiisoft/yii2/issues/12762

Your new loop is fine.